dora/core/client/fs/src/main/java/alluxio/client/file/dora/netty/NettyDataWriter.java

```diff
@@ -0,0 +1,549 @@
+   */
+  private NettyDataWriter(FileSystemContext context, final WorkerNetAddress address,
+                          long length, long packetSize, RequestType type, OutStreamOptions options,
+                          Channel channel) {
+    mContext = context;
+    mAddress = address;
+    mLength = length;
+    Protocol.WriteRequest.Builder builder =
+        Protocol.WriteRequest.newBuilder().setTier(options.getWriteTier())
+            .setType(getRequestType(type));
+
+    Protocol.CreateUfsFileOptions ufsFileOptions =
+        Protocol.CreateUfsFileOptions.newBuilder().setUfsPath(options.getUfsPath())
+            .setOwner(options.getOwner()).setGroup(options.getGroup())
+            .setMode(options.getMode().toShort()).setMountId(options.getMountId()).build();
+    builder.setCreateUfsFileOptions(ufsFileOptions);
+
+    mPartialRequest = builder.buildPartial();
+    mPacketSize = packetSize;
+    mChannel = channel;
+    mChannel.pipeline().addLast(new PacketWriteResponseHandler());
+  }
+
+  @Override
+  public long pos() {
+    try (LockResource lr = new LockResource(mLock)) {
+      return mPosToQueue;
+    }
+  }
+
+  @Override
+  public Optional<String> getUfsContentHash() {
+    return Optional.empty();
+  }
+
+  /**
+   * Write chunk by sending a write request.
+   * @param bytes the bytes to be written
+   * @param off the offset of the given byte array
+   * @param len the length to be written
+   * @throws IOException
+   */
+  public void writeChunk(byte[] bytes, int off, int len) throws IOException {
+    // TODO(JiamingMai): maybe we can reduce copying here
+    ByteBuf byteBuf = mChannel.alloc().buffer(len);
+    byteBuf.writeBytes(bytes, off, len);
+    writeChunk(byteBuf);
+  }
+
+  @Override
+  public void writeChunk(final ByteBuf buf) throws IOException {
+    final long len;
+    final long offset;
+    try (LockResource lr = new LockResource(mLock)) {
+      Preconditions.checkState(!mClosed && !mEOFSent && !mCancelSent);
+      Preconditions.checkArgument(buf.readableBytes() <= mPacketSize);
+      while (true) {
+        if (mPacketWriteException != null) {
+          Throwables.propagateIfPossible(mPacketWriteException, IOException.class);
+          throw AlluxioStatusException.fromCheckedException(mPacketWriteException);
+        }
+        if (!tooManyPacketsInFlight()) {
+          offset = mPosToQueue;
+          mPosToQueue += buf.readableBytes();
+          len = buf.readableBytes();
+          break;
+        }
+        try {
+          if (!mBufferNotFullOrFailed.await(WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+            throw new DeadlineExceededException(
+                String.format("Timeout writing to %s for request %s after %dms.",
+                    mAddress, mPartialRequest, WRITE_TIMEOUT_MS));
+          }
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new CanceledException(e);
+        }
+      }
+    } catch (IOException e) {
```

Comment:
could there be an InterruptedException? Maybe just catch all Throwable?

Reply:
Good catch. Updated.